src/Microsoft.Azure.WebJobs.ServiceBus/Triggers/BrokeredMessageToStringConverter.cs

```diff
@@ -2,54 +2,76 @@
 // Licensed under the MIT License. See License.txt in the project root for license information.
 
 using System;
+using System.Globalization;
 using System.IO;
+using System.Runtime.Serialization;
+using System.Text;
 using System.Threading;
 using System.Threading.Tasks;
+using System.Xml;
 using Microsoft.Azure.WebJobs.Host.Converters;
 using Microsoft.ServiceBus.Messaging;
 
 namespace Microsoft.Azure.WebJobs.ServiceBus.Triggers
 {
     internal class BrokeredMessageToStringConverter : IAsyncConverter<BrokeredMessage, string>
     {
-        public Task<string> ConvertAsync(BrokeredMessage input, CancellationToken cancellationToken)
+        public async Task<string> ConvertAsync(BrokeredMessage input, CancellationToken cancellationToken)
         {
             if (input == null)
             {
                 throw new ArgumentNullException("input");
             }
+            Stream stream = input.GetBody<Stream>();
+            if (stream == null)
+            {
+                return null;
+            }
 
-            if (input.ContentType == ContentTypes.TextPlain ||
-                input.ContentType == ContentTypes.ApplicationOctetStream ||
-                input.ContentType == ContentTypes.ApplicationJson)
+            TextReader reader = new StreamReader(stream, StrictEncodings.Utf8);
+            try
             {
-                Stream stream = input.GetBody<Stream>();
-                if (stream == null)
+                cancellationToken.ThrowIfCancellationRequested();
+                try
                 {
-                    return Task.FromResult<string>(null);
+                    return await reader.ReadToEndAsync();
+                }
+                catch (DecoderFallbackException)
+                {
+                    // we'll try again below
                 }
 
-                try
+                // We may get here if the message is a string yet was DataContract-serialized when created. We'll
+                // try to deserialize it here. This may fail as well, in which case we'll try to give a decent error. We cannot read the 
+                // body twice so we need to use the DataContractSerializer directly rather than use input.GetBody<string>().
+                stream.Position = 0;
+                DataContractSerializer serializer = new DataContractSerializer(typeof(string));
```

Comment:
Can you just create a new local BrokeredMessage passing in the stream to the ctor and use the SB APIs for reading? That way we're still getting their serialization logic.